File: diameter_of_binary_tree.cpp
class Solution {
public:
    int ans = 0;
    int dfs(TreeNode* root) {
        if(!root) return 0;

        int left = dfs(root->left);
        int right = dfs(root->right);

        ans = max(ans, left+right);
        return 1 + max(left, right);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return ans;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(n)

File: diameter_of_binary_tree.go
func dfs(root *TreeNode, ans *int) int {
    if(root == nil) {
        return 0
    }

    left := dfs(root.Left, ans)
    right := dfs(root.Right, ans)

    *ans = max(*ans, left+right)
    return 1 + max(left, right)
}
func diameterOfBinaryTree(root *TreeNode) int {
    ans := 0
    dfs(root, &ans);
    return ans
}

// Time Complexity : O(n)
// Space Complexity : O(n)